### PR TITLE
Adding required flags to non-default arguments in hf_argparser

### LINKED ISSUE #10677: hf_argparser doesn't set the required flag on non-defaulted enums
## Environment info

- `transformers` version: 3.0.0-4.3.3
- Platform: macOS
- Python version: 3.9
- PyTorch version (GPU?): n/a
- Tensorflow version (GPU?): n/a
- Using GPU in script?: n/a
- Using distributed or parallel set-up in script?: n/a

### Who can help

I'm not sure who the owner is of hf_argparser.

## Information

We're using hf_argparser to parse our experiment config into dataclasses before training.

## To reproduce

1. Add an enum argument without a default to a dataclass
2. Parse the command line arguments without supplying the enum argument
3. Should have raised an exception and printed the usage, instead defaults the value to `None`.

## Expected behavior

It should raise an exception. The issue is on https://github.com/huggingface/transformers/blob/master/src/transformers/hf_argparser.py#L100, the if statement should have an else which sets `kwargs["required"]=True`, the same way line [134](https://github.com/huggingface/transformers/blob/master/src/transformers/hf_argparser.py#L134) does.

I can work up a patch if you agree this is an issue. I think it will also occur with anything that falls into [this branch](https://github.com/huggingface/transformers/blob/master/src/transformers/hf_argparser.py#L118) of the if too.

### PR DESCRIPTION
**Pull request by Craigacp**
Signed-off-by: Adam Pocock <adam.pocock@oracle.com>

# What does this PR do?

Fixes #10677. I didn't update the docs as I think this is the intended behaviour, but I can do if you think this change would be unexpected.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Issue #10677.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@sgugger

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for fixing this and adding a test!

**Comment by Craigacp** [comment]
No problem. I can't see what test failed in CircleCI as it wants to bind to my Github account and access private things in my org. Is there any way to see the failures without letting it into my account?

File: src/transformers/hf_argparser.py [hunk 1]
```diff
@@ -124,6 +126,8 @@ def _add_dataclass_arguments(self, dtype: DataClassType):
                 ), "{} cannot be a List of mixed types".format(field.name)
                 if field.default_factory is not dataclasses.MISSING:
                     kwargs["default"] = field.default_factory()
+                else:
+                    kwargs["required"] = True
```
Comment (sgugger) [inline src/transformers/hf_argparser.py:130]: Actually this looks like it ignores the default if present, so it should be
```suggestion
                elif field.default is dataclasses.MISSING:
                    kwargs["required"] = True
```

Comment (Craigacp) [reply src/transformers/hf_argparser.py:130]: Oh yeah, sure.

**Review by LysandreJik — APPROVED** [review]
LGTM! Failing test is flaky and not linked to your PR.

**Pull request merged**